### PR TITLE
Fix Counts object creation with empty input

### LINKED ISSUE #5017: Count constructor crashes when data is empty
### Information

- **Qiskit Terra version**: 0.15.1
- **Python version**: 3.7.3
- **Operating system**: Mac OS Catalina 10.15.6

### What is the current behavior?
I simply ran a circuit which does not perform any measurements. On the returned results, `get_counts()` crashes because an empty dictionary is passed into the `Count` constructor and the constructor tries to iterate on the empty dictionary.

I tried this on an older version, Terra 0.14.2, which instead of `Count` uses `postprocess.format_counts()`, which simply ignores the empty dictionary and returns an empty dictionary in this case. This did not crash.

### Steps to reproduce the problem
Run any circuit without calling `measure()`, then try to apply `get_counts()` to the results.

### What is the expected behavior?
`get_counts()` should return an empty dictionary (and not crash).

### Suggested solutions
A simple check for an empty dictionary in the `Count` constructor would be sufficient, I think.

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

Previously the counts object was not correctly handling when an empty
dict was passed into the constructor. This was because the
postprocessing was unconditionally being called even though it requires
data to work. This commit fixes the issue by checking if we have data
first and if not, just setting all the fields populated by
postprocessing to empty dicts. This is necessary because we
unconditionally build a results object from the get_counts() call even
if there are no measurement results in the data. At the same time error
handling is added for the most_frequent() method to account for the case
when there is no data in the object.

### Details and comments

Fixes #5017

### CODE REVIEW

File: test/python/result/test_counts.py [hunk 1]
```diff
@@ -252,3 +252,43 @@ def test_hex_outcomes_0b_bitstring_counts(self):
...
+        raw_counts = {}
+        counts_obj = counts.Counts(raw_counts)
+        self.assertRaises(exceptions.QiskitError,
+                          counts_obj.most_frequent)
+
+    def test_most_frequent_duplicate_empty_bitstring_counts(self):
+        raw_counts = {}
+        counts_obj = counts.Counts(raw_counts)
+        self.assertRaises(exceptions.QiskitError,
+                          counts_obj.most_frequent)
```
Comment (kdk) [inline test/python/result/test_counts.py:287]: Should this test be duplicated?

**Review by kdk — APPROVED** [review]

File: test/python/result/test_counts.py (hunk 1 above)
Comment (mtreinish) [reply test/python/result/test_counts.py:287]: oops yeah, sloppy copy paste from the other input formats. This was the test case where counts where like `{0: 10, 1: 10}` and checking that properly raises. But when the raw counts are empty that doesn't mean anything. I'll remove this real quickly

Comment (mtreinish) [reply test/python/result/test_counts.py]: Fixed in: https://github.com/Qiskit/qiskit-terra/pull/5018/commits/22d517f4c22c1e159884b105db5b14d4a860d9dd

**Review by kdk — APPROVED** [review]

**Comment by atskae** [comment]
Thank you for the prompt response! Which Terra release will include this fix? Thank you!

**Pull request merged**